docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+   The number of bits required for a pointer varies depending on the pointer representation:
+   - 32-bit integer pointers require 29 bits of state.
+   - 64-bit integer pointers require either 45 or 52 bits, depending on the virtual address space size of the target.
+   - CHERI capabilities require 125 bits of state but provide a 'free' discriminator bit because the tag bit can be used to discriminate CHERI capabilities from non-pointer data.
+ - Value types are still being defined; however, it seems plausible that they will each require one descriptor state and as many payload bits as it has non-padding bits.
+   Making efficient use of padding bits as discriminators seems very difficult and so is likely to be left to a later ABI, if ever implemented.
+
+Once the payload sizes are all known, the algorithm for implementing the packing depends on the type of the largest payload.
+There are two special cases:
+
+If the largest type is a floating-point type and the payload of all other types plus the bits required for the discriminator will fit in the unused bits of the significand of a signalling NaN minus one then the top bit of the significand is reserved to discriminate between a real signalling NaN and another value and the remainder of the types are encoded in the spare bits, starting this algorithm again.
+If the remaining bits of the significand are not sufficient to encode the remaining states then this rule is ignored.
+
+If the largest type is a CHERI capability then the tag bit is used to differentiate pointer and non-pointer types.
+This leaves 128 bits to use for the remainder of the values.
+
+The size of the final object is the number of payload bits, plus the base-two logarithm of the number of discriminator states rounded up.
+The alignment of the result is the alignment of the most strongly aligned type.
+
+If one of the types is a pointer, the object size is pointer sized, and the discriminator is three or fewer bits, then the low bits of the pointer are used for the discriminator.
+Pointer values are stored with their bits in the natural positions and must have the low bits subtracted away on use.
+Other values are stored in the most significant bits.
+
+*Rationale:* Most ISAs make subtracting a 3-bit immediate trivial (often something that can be folded into another instruction).
+Other values that fit in the payload space can be shifted down as required in a single-cycle instruction.
+
+If the payload is an integer number of bytes, the payload is stored first (independent of byte order) so that the payload can benefit from strong alignment, which avoids the payload accidentally spanning a cache line.
+In this case, all elements of the payload are stored starting from the beginning of the payload.
```

Comment:
I'm not sure what you mean by "stored first". Most or least significant bits? Assuming (for now) all types will be power of twos, this paragraph is unnecessary.

Reply:
Right, so beginning is least-significant. Perhaps use that language instead.

Reply:
If you are on a little-endian system.  If you are on a big-endian system, it is the most significant.

Reply:
Ah right, I see, makes sense.